### PR TITLE
DI-240 and DI-241: Bug fix to warm start logic

### PR DESCRIPTION
**Pull request by herlenashavi**
**Changes in PR:**
- Bug fix to warm start logic
- Adding feature analysis to dict representation
- Tests for warm start
- Rebased fixes from upstream.

### CODE REVIEW

File: streamhist/histogram.py [hunk 1]
```diff
@@ -378,35 +393,53 @@ def density(self, p):
             dd = _compute_density(p, bin_i, bin_i1)
         return dd
 
+    def _quantile(self, sums, q):
+        if q <= 0:
+            return self._min
+        if q >= 1:
```
Comment (herlenashavi) [inline streamhist/histogram.py:399]: Pulled from upstream, fixes merged to streamhist.

File: streamhist/histogram.py [hunk 2]
```diff
@@ -159,15 +158,7 @@ def median(self):
...
-        if len(self.bins) >= self.maxbins:
-            # Return the approximate median
-            return self.quantiles(0.5)[0]
-        else:
-            # Return the 'exact' median when possible
-            mid = int(self.total/2)
-            if self.total % 2 == 0:
-                return (self.bins[mid-1] + self.bins[mid]).value
-            return self.bins[mid].value
+        return self.quantiles(0.5)[0]
```
Comment (herlenashavi) [inline streamhist/histogram.py:161]: Pulled from upstream, fixes merged to streamhist.

File: streamhist/histogram.py [hunk 3]
```diff
@@ -39,12 +39,13 @@
...
 
+import math
 import sys
-from bisect import bisect_left
+from bisect import bisect_left, bisect_right
 
 from sortedcontainers import SortedListWithKey
 from .utils import (next_after, bin_diff, accumulate, linspace,
-                    iterator_types, argmin, bin_sums, roots)
+                    iterator_types, argmin, roots)
```
Comment (herlenashavi) [inline streamhist/histogram.py:48]: Pulled from upstream, fixes merged to streamhist.

File: recommended.txt [hunk 4]
```diff
@@ -1,4 +1,5 @@
 # Examples and tests
+pytest
```
Comment (herlenashavi) [inline recommended.txt:2]: Pulled from upstream, fixes merged to streamhist.

File: examples/examples.ipynb [hunk 5]
```diff
@@ -46,11 +46,11 @@
    "source": [
     "# Testing\n",
     "\n",
-    "`streamhist` comes with a relatively comprehensive range of tests, including unit tests and regression tests. To run the tests, you can use `py.test` or `nosetests`, which can both be installed via `pip` using the `recommended.txt` file (note, this will also install `numpy`, `matplotlib`, and `IPython` which are used for tests and examples):\n",
+    "`streamhist` comes with a relatively comprehensive range of tests, including unit tests and regression tests. To run the tests, you can use `pytest`, which can be installed via `pip` using the `recommended.txt` file (note, this will also install `numpy`, `matplotlib`, and `IPython` which are used for tests and examples):\n",
```
Comment (herlenashavi) [inline examples/examples.ipynb:49]: Pulled from upstream, fixes merged to streamhist.

File: docs/examples.rst [hunk 6]
```diff
@@ -48,15 +48,15 @@ Testing
...
+``pytest``, which can be installed via ``pip``
 using the ``recommended.txt`` file (note, this will also install
 ``numpy``, ``matplotlib``, and ``IPython`` which are used for tests and
 examples):
 
 .. code:: bash
 
     pip install -r recommended.txt
-    nosetests streamhist
+    pytest streamhist
```
Comment (herlenashavi) [inline docs/examples.rst:59]: Pulled from upstream, fixes merged to streamhist.

File: README.rst [hunk 7]
```diff
@@ -48,15 +49,15 @@ Testing
 
 ``streamhist`` comes with a relatively comprehensive range of tests,
 including unit tests and regression tests. To run the tests, you can use
-``py.test`` or ``nosetests``, which can both be installed via ``pip``
+``pytest``, which can be installed via ``pip``
```
Comment (herlenashavi) [inline README.rst:52]: Pulled from upstream, fixes merged to streamhist.

File: streamhist/histogram.py [hunk 8]
```diff
@@ -378,35 +393,53 @@ def density(self, p):
...
-                else:
-                    bin_i1 = self.bins[index]
-                if index:
-                    prev_sum = sums[index-1]
-                else:
-                    prev_sum = 0.0
-                qq = _compute_quantile(target_sum, bin_i, bin_i1, prev_sum+1)
-            result.append(qq)
-        return result
+        # Deviation from Ben-Haim sum strategy:
```
Comment (herlenashavi) [inline streamhist/histogram.py:434]: Pulled from upstream, fixes merged to streamhist.

File: streamhist/test/test_histogram.py [hunk 9]
```diff
@@ -304,6 +306,54 @@ def test_quantiles():
     assert about(c, 0.66, 0.05)
 
 
+def test_histogram_exact():
```
Comment (herlenashavi) [inline streamhist/test/test_histogram.py:309]: Pulled from upstream, fixes merged to streamhist.

File: streamhist/test/test_histogram.py [hunk 10]
```diff
@@ -304,6 +306,54 @@ def test_quantiles():
...
+            np.quantile(points, [0] * 7 + q),
+            np.quantile(points, q[7:] + [1] * 7)):
+        err_denom = b_np_max - b_np_min
+        err_sum += abs(b - b_np) / err_denom
+    assert err_sum <= expected_error
+    assert h.mean() == approx(np.mean(points))
+    assert h.var() == approx(np.var(points), rel=.05)
+    assert h.min() == min(points)
+    assert h.max() == max(points)
+    assert h.count() == num_points
```
Comment (herlenashavi) [inline streamhist/test/test_histogram.py:354]: Pulled from upstream, fixes merged to streamhist.

File: streamhist/utils.py [hunk 11]
```diff
@@ -34,7 +34,7 @@ def accumulate(iterable):
 
 _E = 2.718281828459045
 
-__all__ = ["next_after", "bin_sums", "argmin", "bin_diff", "accumulate"]
+__all__ = ["next_after", "argmin", "bin_diff", "accumulate"]
```
Comment (herlenashavi) [inline streamhist/utils.py:37]: Pulled from upstream, fixes merged to streamhist.

**Review by avinashkz — APPROVED** [review]

**Pull request merged**